codec/reflectcodec/type_codec.go

```diff
@@ -332,6 +352,64 @@ func (c *genericCodec) marshal(value reflect.Value, p *wrappers.Packer, maxSlice
 				return err
 			}
 		}
+		return nil
+	case reflect.Map:
+		keys := value.MapKeys()
+		numElts := len(keys)
+		if uint32(numElts) > maxSliceLen {
+			return fmt.Errorf("%w; map length, %d, exceeds maximum length, %d",
+				codec.ErrMaxSliceLenExceeded,
+				numElts,
+				maxSliceLen,
+			)
+		}
+		p.PackInt(uint32(numElts)) // pack # elements
+		if p.Err != nil {
+			return p.Err
+		}
+
+		// pack key-value pairs sorted by increasing key
+		type keyTuple struct {
+			key   reflect.Value
+			bytes []byte
+		}
+
+		sortedKeys := make([]keyTuple, len(keys))
+		for i, key := range keys {
+			// Note this [p] shadows the outer [p]
+			p := wrappers.Packer{
+				MaxSize: int(c.maxSliceLen),
+				Bytes:   make([]byte, 0, 128),
+			}
+			if err := c.marshal(key, &p, c.maxSliceLen); err != nil {
+				return err
+			}
+			if p.Err != nil {
+				return fmt.Errorf("couldn't marshal map key %+v: %w ", key, p.Err)
+			}
+			sortedKeys[i] = keyTuple{
+				key:   key,
+				bytes: p.Bytes,
+			}
+		}
```

Comment:
This feels pretty sub-optimal... Could we add a TODO to remove the usage of the inner packer? We could just track the offsets to get the bytes. (Without needing to allocate new packers and bytes every time). I think we'd probably still need one allocation to avoid overwriting the keys when packing the values (and we'd definitely want to avoid re-marshalling the keys over avoiding the single memory allocation).

Reply:
If we changed the format to:
```
Length
Key0
Key1
...
Value0
Value1
...
```
I'm wondering if we could fully remove the additional allocation.... I don't think we can... but I felt like it was an interesting thought.